File: ios/HttpClient.swift
import Foundation

struct RelayDeliveryResult {
  let success: Bool
  let statusCode: Int?
  let errorMessage: String?
}

enum HttpClient {
  static func post(
    endpoint: String,
    headersJson: String?,
    timeoutMs: Double?,
    body: Data
  ) -> RelayDeliveryResult {
    let bodyString = String(data: body, encoding: .utf8) ?? "<binary body>"

    guard let url = URL(string: endpoint) else {
      RelayLogger.error("Invalid endpoint URL: \(endpoint)")
      return RelayDeliveryResult(
        success: false,
        statusCode: nil,
        errorMessage: "Invalid endpoint URL."
      )
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.httpBody = body
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")

    if let timeoutMs, timeoutMs > 0 {
      request.timeoutInterval = timeoutMs / 1000
    }

    if let headersJson, let headersData = headersJson.data(using: .utf8),
       let headersObject = try? JSONSerialization.jsonObject(with: headersData) as? [String: String]
    {
      for (key, value) in headersObject {
        request.setValue(value, forHTTPHeaderField: key)
      }
    }

    let requestHeaders = request.allHTTPHeaderFields ?? [:]
    RelayLogger.info("POST \(endpoint)")
    RelayLogger.info("Request headers: \(Self.formatHeaders(requestHeaders))")
    RelayLogger.info("Request body: \(bodyString)")

    let semaphore = DispatchSemaphore(value: 0)
    var result = RelayDeliveryResult(
      success: false,
      statusCode: nil,
      errorMessage: "Request did not complete."
    )

    let task = URLSession.shared.dataTask(with: request) { data, response, error in
      defer { semaphore.signal() }

      if let error {
        RelayLogger.error("Request failed: \(error.localizedDescription)")
        result = RelayDeliveryResult(
          success: false,
          statusCode: nil,
          errorMessage: error.localizedDescription
        )
        return
      }

      let statusCode = (response as? HTTPURLResponse)?.statusCode
      let responseBody = data.flatMap { String(data: $0, encoding: .utf8) } ?? "<empty>"
      let success = statusCode.map { (200 ... 299).contains($0) } ?? false

      if success {
        RelayLogger.info("Response status: \(statusCode ?? 0)")
        RelayLogger.info("Response body: \(responseBody)")
      } else {
        RelayLogger.error("Response status: \(statusCode ?? 0)")
        RelayLogger.error("Response body: \(responseBody)")
      }

      result = RelayDeliveryResult(
        success: success,
        statusCode: statusCode,
        errorMessage: success ? nil : "HTTP \(statusCode ?? 0)"
      )
    }

    task.resume()
    semaphore.wait()

    return result
  }

  private static func formatHeaders(_ headers: [String: String]) -> String {
    guard let data = try? JSONSerialization.data(withJSONObject: headers, options: [.sortedKeys]),
          let json = String(data: data, encoding: .utf8)
    else {
      return "{}"
    }
    return json
  }
}

File: android/src/main/java/com/margelo/nitro/backgroundlocationrelay/HttpClient.kt
package com.margelo.nitro.backgroundlocationrelay

import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONObject

data class RelayDeliveryResult(
  val success: Boolean,
  val statusCode: Int?,
  val errorMessage: String?,
)

object HttpClient {
  fun post(
    endpoint: String,
    headersJson: String?,
    timeoutMs: Double?,
    body: String,
  ): RelayDeliveryResult {
    return try {
      val connection = URL(endpoint).openConnection() as HttpURLConnection
      val timeout = timeoutMs?.toInt()?.takeIf { it > 0 } ?: 15_000

      connection.requestMethod = "POST"
      connection.doOutput = true
      connection.connectTimeout = timeout
      connection.readTimeout = timeout
      connection.setRequestProperty("Content-Type", "application/json")

      if (!headersJson.isNullOrEmpty()) {
        val headers = JSONObject(headersJson)
        headers.keys().forEach { key ->
          connection.setRequestProperty(key, headers.getString(key))
        }
      }

      val requestHeaders = connection.requestProperties
      RelayLogger.info("POST $endpoint")
      RelayLogger.info("Request headers: ${formatHeaders(requestHeaders)}")
      RelayLogger.info("Request body: $body")

      connection.outputStream.use { stream ->
        stream.write(body.toByteArray(Charsets.UTF_8))
      }

      val statusCode = connection.responseCode
      val responseBody = readResponseBody(connection)
      connection.disconnect()

      val success = statusCode in 200..299
      if (success) {
        RelayLogger.info("Response status: $statusCode")
        RelayLogger.info("Response body: $responseBody")
      } else {
        RelayLogger.error("Response status: $statusCode")
        RelayLogger.error("Response body: $responseBody")
      }

      RelayDeliveryResult(
        success = success,
        statusCode = statusCode,
        errorMessage = if (success) null else "HTTP $statusCode",
      )
    } catch (error: Exception) {
      RelayLogger.error("Request failed: ${error.message ?: "unknown error"}")
      RelayDeliveryResult(
        success = false,
        statusCode = null,
        errorMessage = error.message ?: "Request failed.",
      )
    }
  }

  private fun readResponseBody(connection: HttpURLConnection): String {
    val stream =
      if (connection.responseCode in 200..299) {
        connection.inputStream
      } else {
        connection.errorStream
      }

    return stream?.bufferedReader()?.use { it.readText() }?.ifEmpty { "<empty>" } ?: "<empty>"
  }

  private fun formatHeaders(headers: Map<String, List<String>>): String {
    val json = JSONObject()
    headers.forEach { (key, values) ->
      if (key.isNotEmpty() && values.isNotEmpty()) {
        json.put(key, values.joinToString(", "))
      }
    }
    return json.toString()
  }
}
